pkg/services/ngalert/notifier/alertmanager.go

```diff
@@ -260,61 +200,48 @@ func (am *Alertmanager) Ready() bool {
 	// We consider AM as ready only when the config has been
 	// applied at least once successfully. Until then, some objects
 	// can still be nil.
-	am.reloadConfigMtx.RLock()
-	defer am.reloadConfigMtx.RUnlock()
-
-	return am.ready()
+	return am.Base.Ready()
 }
 
 func (am *Alertmanager) ready() bool {
 	return am.config != nil
 }
 
 func (am *Alertmanager) StopAndWait() {
-	if am.dispatcher != nil {
-		am.dispatcher.Stop()
-	}
-
-	if am.inhibitor != nil {
-		am.inhibitor.Stop()
-	}
-
-	am.alerts.Close()
-
-	close(am.stopc)
-
-	am.wg.Wait()
+	am.Base.StopAndWait()
 }
 
 // SaveAndApplyDefaultConfig saves the default configuration the database and applies the configuration to the Alertmanager.
 // It rollbacks the save if we fail to apply the configuration.
 func (am *Alertmanager) SaveAndApplyDefaultConfig(ctx context.Context) error {
-	am.reloadConfigMtx.Lock()
```

Comment:
The diff here is ruined because we need to take `Base`'s lock, which causes nesting - there's no functional change here. We just need to make sure that we lock (with the same lock) all the way down to calling `applyConfig` which does not use Base's lock.

Reply:
It preserved it well enough to follow. These lock changes seem straightforward :+1: